src/community/ogcapi/ogcapi-core/src/main/java/org/geoserver/ogcapi/AbstractHTMLMessageConverter.java

```diff
@@ -98,52 +100,69 @@ protected HashMap<String, Object> setupModel(Object value) {
-                                        baseURL,
-                                        (String) unwrapArgument(arguments.get(0)),
-                                        null,
-                                        URLMangler.URLType.RESOURCE));
-        model.put(
-                "externalLink",
-                (TemplateMethodModelEx)
-                        arguments ->
-                                ResponseUtils.buildURL(
-                                        baseURL,
-                                        (String) unwrapArgument(arguments.get(0)),
-                                        null,
-                                        URLMangler.URLType.EXTERNAL));
+                        arguments -> simpleLinkFunction(baseURL, arguments, RESOURCE));
         model.put(
                 "htmlExtensions",
                 (TemplateMethodModelEx)
-                        arguments -> {
-                            if (arguments != null) {
-                                arguments = unwrapArguments(arguments);
-                            }
-                            return processHtmlExtensions(model, arguments);
-                        });
+                        arguments -> processHtmlExtensions(model, unwrapArguments(arguments)));
         model.put("loadJSON", parseJSON());
     }
 
+    private String simpleLinkFunction(String baseURL, List arguments, URLMangler.URLType urlType) {
+        return ResponseUtils.buildURL(
+                baseURL, (String) unwrapArgument(arguments.get(0)), null, urlType);
+    }
+
+    /** Builds a service link back to the same service. Used for backlinks. */
+    private String serviceLink(List arguments) {
+        APIRequestInfo requestInfo = APIRequestInfo.get();
+        return ResponseUtils.buildURL(
+                requestInfo.getBaseURL(),
+                ResponseUtils.appendPath(
+                        requestInfo.getServiceLandingPage(),
+                        (String) unwrapArgument(arguments.get(0))),
+                arguments.size() > 1
+                        ? Collections.singletonMap("f", (String) unwrapArgument(arguments.get(1)))
+                        : null,
+                URLMangler.URLType.SERVICE);
+    }
+
+    /** Builds a service link with the provided path, does not inject the current service path */
+    private String genericServiceLink(List arguments) {
+        APIRequestInfo requestInfo = APIRequestInfo.get();
+        return ResponseUtils.buildURL(
+                requestInfo.getBaseURL(),
+                (String) unwrapArgument(arguments.get(0)),
+                arguments.size() > 1
+                        ? argumentsToKVP(arguments.subList(1, arguments.size()))
+                        : null,
+                URLMangler.URLType.SERVICE);
+    }
+
+    @SuppressWarnings("unchecked")
+    private Map<String, String> argumentsToKVP(List arguments) {
+        if (arguments.size() % 2 != 0)
```

Comment:
If the first argument is followed by n*pairs, shouldn't this be arguments.size() % 2 != 1  ?

Reply:
The method has been given a sublist, skipping the first argument.

Reply:
Which I could make clearer by calling the argument "kvp" rather than "arguments"